atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/InstrumentedCassandraClient.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.cassandra.thrift.Mutation;
+import org.apache.cassandra.thrift.SlicePredicate;
+import org.apache.cassandra.thrift.TimedOutException;
+import org.apache.cassandra.thrift.UnavailableException;
+import org.apache.thrift.TException;
+
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.atlasdb.tracing.CloseableTrace;
+import com.palantir.atlasdb.util.AtlasDbMetrics;
+import com.palantir.processors.AutoDelegate;
+
+@AutoDelegate(typeToExtend = Cassandra.Client.class)
+@SuppressWarnings({"all"}) // thrift variable names
+public class InstrumentedCassandraClient extends AutoDelegate_Client {
+    private static final String SERVICE_NAME = "cassandra-thrift-client";
+    private static final MetricRegistry METRIC_REGISTRY = AtlasDbMetrics.getMetricRegistry();
+
+    private static final Timer MULTIGET_SLICE_TIMER = METRIC_REGISTRY.timer(
+            MetricRegistry.name(InstrumentedCassandraClient.class, "multiget_slice"));
+    private static final Timer BATCH_MUTATE_TIMER = METRIC_REGISTRY.timer(
+            MetricRegistry.name(InstrumentedCassandraClient.class, "batch_mutate"));
+    private static final Timer GET_RANGE_SLICE_TIMER = METRIC_REGISTRY.timer(
+            MetricRegistry.name(InstrumentedCassandraClient.class, "get_range"));
+
+    private Cassandra.Client delegate;
+
+    public InstrumentedCassandraClient(Cassandra.Client delegate) {
+        super(delegate.getInputProtocol());
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Cassandra.Client delegate() {
+        return delegate;
+    }
+
+    @Override
+    public void batch_mutate(Map<ByteBuffer, Map<String, List<Mutation>>> mutation_map,
+            ConsistencyLevel consistency_level)
+            throws InvalidRequestException, UnavailableException, TimedOutException, TException {
+        //noinspection unused - try-with-resources closes trace
+        try (CloseableTrace trace = startLocalTrace("client.batch_mutate(number of rows {}, consistency {})",
+                mutation_map.size(), toString(consistency_level))) {
+            registerDuration(() -> {
+                delegate.batch_mutate(mutation_map, consistency_level);
+                return null;
+            }, BATCH_MUTATE_TIMER);
+        }
+    }
+
+    @Override
+    public Map<ByteBuffer, List<ColumnOrSuperColumn>> multiget_slice(List<ByteBuffer> keys, ColumnParent column_parent,
+            SlicePredicate predicate, ConsistencyLevel consistency_level)
+            throws InvalidRequestException, UnavailableException, TimedOutException, TException {
+        //noinspection unused - try-with-resources closes trace
+        try (CloseableTrace trace = startLocalTrace(
+                "client.multiget_slice(table {}, number of keys {}, number of columns {}, consistency {})",
+                LoggingArgs.safeInternalTableNameOrPlaceholder(column_parent.column_family),
```

Comment:
i guess tracing doesn't support `SafeArg`?

Reply:
Yep. We just do MessageFormatter.arrayFormat().getMessage() at the end.